### ISSUE TITLE
Nginx Unit not conforming to ASGI spec on websocket connections

### ISSUE THREAD

**Issue by FixFlare** [nginx/unit#1507]
I'm using the "Litestar" package to host a python API using Nginx Unit.

During development I used Litestar's recommended dev server "uvicorn" and did not experience any issues until I deployed to production running Nginx Unit and then got the error "405: Method Not Allowed" when I tried to make my websocket connections.

Tracing the issue, I found that the method being sent on the websocket connection was set to "GET" and the way Litestar handles determining if it's a websocket connection is based on the method being "None" so the fact that Nginx Unit is sending "GET" for the request is breaking it.

When I check the [websocket spec](https://datatracker.ietf.org/doc/html/rfc6455), page 17, point 2 it says

"The method of the request MUST be GET, and the HTTP version MUST be at least 1.1."

So I initially thought Litestar was the problem and filed an issue over there.

The Litestar maintainers advised that the ASGI spec does not include a "method" key and so it would in theory be against the ASGI spec for Nginx Unit to be sending it.

> If you take a look at the ASGI specs, the `websocket` scope does not define a `method` key: https://asgi.readthedocs.io/en/latest/specs/www.html#websocket-connection-scope.
> 
> The RFC you are referring to is the initial request, which isn't handled by the ASGI app, but the ASGI server.

I have a more detailed writeup over on the Litestar github [here](https://github.com/litestar-org/litestar/issues/3887)

I also made a git repo that allows you to reproduce the problem here:  [https://github.com/FixFlare/bug_3887](https://github.com/FixFlare/bug_3887)

If I can clarify anything about the repo or the issue please let me know.

Thanks,

**Comment by gourav-kandoria** [comment]
@FixFlare My say in this would be that deducing the connection type should not be done from the "method" field of the scope object.
"type" field  is there for that. I agree, "method" field is not relevant in websocket scope object. But in that case, framework shouldn't be bothered about it whether it is present or not.
BTW, can't comment directly whether it is to be considered as breaking specification from unit's end or not. In this case, it might be seen as breaking specification. But in other cases, where framework don't bother about this field. things should work correctly

Although, If this need handling at nginx unit end. can contribute to that.
@callahad @ac000

**Comment by provinzkraut** [comment]
> My say in this would be that deducing the connection type should not be done from the "method" field of the scope object.
"type" field is there for that. I agree, "method" field is not relevant in websocket scope object. But in that case, framework shouldn't be bothered about it whether it is present or not.

As a framework author, I'm not sure I disagree with this. While it may not be *ideal* to make assumptions based on the method, it's certainly *valid* according to the spec, and I would assume this to work with every server implementing the spec correctly.

We could of course change the behaviour in Litestar, but that doesn't prevent other frameworks from having the same issues when deployed with nginx unit, if they rely on the spec being implemented the same way we do, so I believe fixing this here would be more beneficial overall.

**Comment by FixFlare** [comment]
@gourav-kandoria 

I understand the point being made by both you and @provinzkraut.


As a neutral party I think it is a slippery slope if we say that you can take any specification and start adding your own fields as long as they don't already exist in the spec.

What if the spec decides to officially add that field at a later date?

IMO, I think that would set a bad precedent for the open source community.

**Comment by ac000** [comment]
Hey folks.

[Trying to reproduce this...]

I found something which still has redis in its repositories (aioredis doesn't seem to like valkey...)

So I installed redis and a compiled Unit from master with python support and the app from <https://github.com/FixFlare/bug_3887> on a RHEL 9 VM, with the following unit config

```json
 {
    "listeners": {
        "[::]:8080": {
            "pass": "routes/bug_3887"
        }
    },

    "routes": {
        "bug_3887": [
            {
                "match": {
                    "host": "*",
                    "uri": [
                        "/api/*",
                        "/api"
                    ]
                },
                "action": {
                    "pass": "applications/bug_3887-server"
                }
            }, {
                "match": {
                    "host": "*"
                },
                "action": {
                    "share": "/home/andrew/src/python/gh1507/public$uri",
                    "fallback": {
                        "share": "/home/andrew/src/python/gh1507/public/index.html"
                    }
                }
            }
        ]
    },

    "applications": {
        "bug_3887-server": {
            "type": "python",
            "path": "/home/andrew/src/python/gh1507/",
            "module": "app",
            "callable": "app"
        }
    }
}
```

I made a tweak to the index.html for the websocket host and port, I.e.

```diff
diff --git a/bug_3887/src/bug_3887/public/index.html b/bug_3887/src/bug_3887/pub
lic/index.html
index 0f349d9..80846a0 100644
--- a/bug_3887/src/bug_3887/public/index.html
+++ b/bug_3887/src/bug_3887/public/index.html
@@ -29,11 +29,11 @@
   </style>
 </head>
 <body>
-  <button id="connectButton">Connect WebSocket Port 80 to Nginx Unit</button>
+  <button id="connectButton">Connect WebSocket Port 8080 to Nginx Unit</button>
 
   <script>
     document.getElementById('connectButton').addEventListener('click', () => {
-      const url = "ws://localhost:80/api/ws/1234"; // Replace with your WebSocket URL
+      const url = "ws://rhel-9:8080/api/ws/1234"; // Replace with your WebSocket URL
       const socket = new WebSocket(url);
 
     
```

Then from my Fedora workstation hit up `http://rhel-9:8080/index.html` in Firefox. I get the page with the blue button. But clicking it doesn't do anything as it can't seem to connect to the websocket server.

From the firefox console...

```
GET ws://rhel-9:8080/api/ws/1234                                                                            NS_ERROR_WEBSOCKET_CONNECTION_REFUSED

Firefox can’t establish a connection to the server at ws://rhel-9:8080/api/ws/1234. [index.html:37:22](http://rhel-9:8080/index.html)
WebSocket error: 
error { target: WebSocket, isTrusted: true, srcElement: WebSocket, currentTarget: WebSocket, eventPhase: 2, bubbles: false, cancelable: false, returnValue: true, defaultPrevented: false, composed: false, … }
```

I tested with my own websocket server running on the VM connecting to it from Firefox on the WS via `ws://rhel-9:1975/` which works fine, so it isn't a Firefox won't connect to non-TLS remote websocket issue...

The app is running

```
2025/01/22 16:21:53 [info] 74547#74547 router started
2025/01/22 16:21:53 [info] 74548#74548 "bug_3887-server" prototype started
2025/01/22 16:21:53 [info] 74549#74549 "bug_3887-server" application started
```

Any idea's?

Is the websocket path (`/api/ws/1234`) correct?

**Comment by ac000** [comment]
I'm getting this back when it fails

```
HTTP/1.1 405 Method Not Allowed
content-type: application/json
content-length: 49
access-control-allow-origin: *
access-control-allow-headers: *
access-control-allow-methods: *
Server: Unit/1.35.0
Date: Wed, 22 Jan 2025 20:21:37 GMT
```

There are no errors in the unit log.

**Comment by gourav-kandoria** [comment]
@ac000 , This is coming from the python application running.
So, this is happened because, for python to work with servers. If follows this specification
https://asgi.readthedocs.io/en/latest/specs/www.html#websocket-connection-scope

So, here unit have passed everything to python app for setting up websocket connection with the python app. But 
an extra field("method") which should be there when setting up http connect but not when setting up websocket connection.

Now, python framework is inferring connection type from this method field and handling it like http request. and in the 
app. for that path there is not handler defined . That is why framework is returning this as an error.

So, basically, cut short:

1) Framework treating it like http connection. and sending response headers with 405 because, no path was there.(just curious how would things would have been played out if a path was defined there)
2) unit assuming , the websocket connection was not accepted by the application and sending back the same to client.

@provinzkraut @FixFlare right? am I missing anything here.

**Comment by ac000** [comment]
Thanks @gourav-kandoria I think I understand now...

When we create a new scope we set

```
nxt_py_asgi_create_http_scope: dict [scope] key [http_version] val [1.1]
nxt_py_asgi_create_http_scope: dict [scope] key [scheme] val [http]
nxt_py_asgi_create_http_scope: dict [scope] key [method] val [GET]
nxt_py_asgi_create_http_scope: dict [scope] key [path] val [/api/ws/1234]
nxt_py_asgi_create_http_scope: dict [scope] key [raw_path] val [(null)]
nxt_py_asgi_create_http_scope: dict [scope] key [query_string] val [(null)]
nxt_py_asgi_create_http_scope: dict [scope] key [client] val [(null)]
nxt_py_asgi_create_http_scope: dict [scope] key [server] val [(null)]
nxt_py_asgi_create_http_scope: dict [scope] key [headers] val [(null)]
```

Regardless if it's for http(s):// or ws(s)://

**Comment by gourav-kandoria** [comment]
> Thanks [@gourav-kandoria](https://github.com/gourav-kandoria) I think I understand now...
> 
> When we create a new scope we set
> 
> ```
> nxt_py_asgi_create_http_scope: dict [scope] key [http_version] val [1.1]
> nxt_py_asgi_create_http_scope: dict [scope] key [scheme] val [http]
> nxt_py_asgi_create_http_scope: dict [scope] key [method] val [GET]
> nxt_py_asgi_create_http_scope: dict [scope] key [path] val [/api/ws/1234]
> nxt_py_asgi_create_http_scope: dict [scope] key [raw_path] val [(null)]
> nxt_py_asgi_create_http_scope: dict [scope] key [query_string] val [(null)]
> nxt_py_asgi_create_http_scope: dict [scope] key [client] val [(null)]
> nxt_py_asgi_create_http_scope: dict [scope] key [server] val [(null)]
> nxt_py_asgi_create_http_scope: dict [scope] key [headers] val [(null)]
> ```
> 
> Regardless if it's for http(s):// or ws(s)://

@ac000 yes, exactly  setting this field only for http and not setting for ws would make unit strictly conform to ASGI spec

**Comment by ac000** [comment]
If I hack it so that the fact we're a websocket flows through to python (and don't set the method in such cases) we get

```
nxt_py_asgi_create_http_scope: dict [scope] key [http_version] val [1.1]
nxt_py_asgi_create_http_scope: dict [scope] key [scheme] val [ws]
nxt_py_asgi_create_http_scope: dict [scope] key [path] val [/api/ws/1234]
nxt_py_asgi_create_http_scope: dict [scope] key [raw_path] val [(null)]
nxt_py_asgi_create_http_scope: dict [scope] key [query_string] val [(null)]
nxt_py_asgi_create_http_scope: dict [scope] key [client] val [(null)]
nxt_py_asgi_create_http_scope: dict [scope] key [server] val [(null)]
nxt_py_asgi_create_http_scope: dict [scope] key [headers] val [(null)]
```

So no method and the scheme now shows as ws... however this just results in a 503 Service Unavailable.

**Comment by ac000** [comment]
What works and what doesn't (as of Unit master (clean)).

There are a bunch of ASGI WebSocket pytests that get run and all pass.

I found a simple FastAPI websocket app

```python
from fastapi import FastAPI, WebSocket

app = FastAPI()

@app.websocket("/ws")
async def websocket_endpoint(websocket: WebSocket):
    print("Listening on WebSocket")
    await websocket.accept()
    while True:
        data = await websocket.receive_text()
        await websocket.send_text(f"Message text was: {data}")
```

Running that under Unit with

```json
{
    "listeners": {
        "[::]:8000": {
            "pass": "applications/fapi-ws"
        }
    },

    "applications": {
        "fapi-ws": {
            "type": "python",
            "path": "/home/andrew/src/python/",
            "module": "fapi-ws",
            "callable": "app"
        }
    }
}
```

Then loading the accompanying web page into Firefox (straight off the filesystem)

```html
<!DOCTYPE html>
<html>
    <head>
        <title>Chat</title>
    </head>
    <body>
        <h1>WebSocket Chat</h1>
        <form action="" onsubmit="sendMessage(event)">
            <input type="text" id="messageText" autocomplete="off"/>
            <button>Send</button>
        </form>
        <ul id='messages'>
        </ul>
        <script>
            var ws = new WebSocket("ws://rhel-9:8000/ws");
            ws.onmessage = function(event) {
                var messages = document.getElementById('messages')
                var message = document.createElement('li')
                var content = document.createTextNode(event.data)
                message.appendChild(content)
                messages.appendChild(message)
            };
            function sendMessage(event) {
                var input = document.getElementById("messageText")
                ws.send(input.value)
                input.value = ''
                event.preventDefault()
            }
        </script>
    </body>
</html>
```

Under Unit this results in 

From Firefox
```
GET /ws HTTP/1.1
Host: rhel-9:8000
User-Agent: Mozilla/5.0 (X11; Linux x86_64; rv:134.0) Gecko/20100101 Firefox/134.0
Accept: */*
Accept-Language: en-GB,en;q=0.5
Accept-Encoding: gzip, deflate
Sec-WebSocket-Version: 13
Origin: null
Sec-WebSocket-Extensions: permessage-deflate
Sec-WebSocket-Key: RssVkEWPk1X6erJYPszzFg==
DNT: 1
Connection: keep-alive, Upgrade
Pragma: no-cache
Cache-Control: no-cache
Upgrade: websocket
```

From Unit
```
HTTP/1.1 404 Not Found
content-length: 22
content-type: application/json
Server: Unit/1.35.0
Date: Thu, 23 Jan 2025 23:32:19 GMT

{"detail":"Not Found"}
```

(Not sure what it didn't find...)

However that does work under uvicorn.

It **does** work under **Unit**  when using the wsdump client from [websocket-client](https://github.com/websocket-client/websocket-client).

Trying to run one of the Unit pytest ASGI websocket apps under Unit

```python
async def application(scope, receive, send):
    if scope['type'] == 'websocket':
        print("WebSocket!")
        while True:
            m = await receive()
            if m['type'] == 'websocket.connect':
                print("Waiting for websocket.accept")
                await send({'type': 'websocket.accept'})
                print("Got WebSocket connection")

            if m['type'] == 'websocket.receive':
                print("Sending response");
                await send(
                    {
                        'type': 'websocket.send',
                        'bytes': m.get('bytes', None),
                        'text': m.get('text', None),
                    }
                )

            if m['type'] == 'websocket.disconnect':
                break
```

Again it fails when trying to access it from Firefox using the above webpage.

```
GET /ws HTTP/1.1
Host: rhel-9:8000
User-Agent: Mozilla/5.0 (X11; Linux x86_64; rv:134.0) Gecko/20100101 Firefox/134.0
Accept: */*
Accept-Language: en-GB,en;q=0.5
Accept-Encoding: gzip, deflate
Sec-WebSocket-Version: 13
Origin: null
Sec-WebSocket-Extensions: permessage-deflate
Sec-WebSocket-Key: ckhEnA8kgUUzH43nqi2K8A==
DNT: 1
Connection: keep-alive, Upgrade
Pragma: no-cache
Cache-Control: no-cache
Upgrade: websocket
```

Returns

```
HTTP/1.1 200 OK
Content-Type: text/plain
Server: Unit/1.35.0
Date: Fri, 24 Jan 2025 00:03:58 GMT
Transfer-Encoding: chunked

0
```

Not the (that you do get when using wsdump, which does work)

```
HTTP/1.1 101 Switching Protocols
Server: Unit/1.35.0
Date: Fri, 24 Jan 2025 00:05:22 GMT
Upgrade: websocket
Connection: Upgrade
Sec-WebSocket-Accept: EAKgOODccXbEgpsJsXHCdePe4fI=
```

you'd be hoping for...

All a bit of a mystery currently...

Hmm, just had a thought to try Chromium... and it works...

```
GET /ws HTTP/1.1
Host: rhel-9:8000
Connection: Upgrade
Pragma: no-cache
Cache-Control: no-cache
User-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/132.0.0.0 Safari/537.36
Upgrade: websocket
Origin: null
Sec-WebSocket-Version: 13
Accept-Encoding: gzip, deflate
Accept-Language: en-GB,en-US;q=0.9,en;q=0.8
Sec-WebSocket-Key: J0XmBYrXNHNQlLng+TW5QA==
Sec-WebSocket-Extensions: permessage-deflate; client_max_window_bits
```

```
HTTP/1.1 101 Switching Protocols
Server: Unit/1.35.0
Date: Fri, 24 Jan 2025 00:20:31 GMT
Upgrade: websocket
Connection: Upgrade
Sec-WebSocket-Accept: 2mU9tHm1Zu0y9K7/6oimM7m0bq4=
```

So colour me extra confused!

**Comment by ac000** [comment]
Another difference, with Chromium the Litestar app gives me a 503 Service Unavailable (as opposed to the 405 Method Not Allowed) with Firefox).

**Comment by gourav-kandoria** [comment]
The difference I see in the request from firefox and chromium is in the field ("Connection")

Chromium: Connection: Upgrade
Firefox: Connection: keep-alive, Upgrade

So, my guess is in case of firefox, unit might have not have been able to guess that it is websocket upgrade request. (not sure exactly can be verified by printing scope object in the python app itself)


Also, For  case where unit is able to identify it is websocket upgrade request. then I guess changing line at this places might do the work in the nxt_py_asgi_create_http_scope function, instead of setting method for both type(http and ws). It can be set conditionaly 
by checking the type value.
 
SET_ITEM(scope, method, v)

**Comment by gourav-kandoria** [comment]
> The difference I see in the request from firefox and chromium is in the field ("Connection")
> 
> Chromium: Connection: Upgrade Firefox: Connection: keep-alive, Upgrade
> 
> So, my guess is in case of firefox, unit might have not have been able to guess that it is websocket upgrade request. (not sure exactly can be verified by printing scope object in the python app itself)
> 
> Also, For case where unit is able to identify it is websocket upgrade request. then I guess changing line at this places will make unit strictly asgi conformant in the nxt_py_asgi_create_http_scope function, instead of setting method for both type(http and ws). It can be set conditionaly by checking the type value.
> 
> SET_ITEM(scope, method, v)

just checked the same with this python request and indeed this is happening(try both variation of fields)

```
import requests

headers = {
    "User-Agent": "Mozilla/5.0 (X11; Linux x86_64; rv:134.0) Gecko/20100101 Firefox/134.0",
    "Accept": "*/*",
    "Accept-Language": "en-GB,en;q=0.5",
    "Accept-Encoding": "gzip, deflate",
    "Sec-WebSocket-Version": "13",
    "Origin": "null",
    "Sec-WebSocket-Extensions": "permessage-deflate",
    "Sec-WebSocket-Key": "ckhEnA8kgUUzH43nqi2K8A==",
    "DNT": "1",
    "Connection": "keep-alive, Upgrade",
    "Pragma": "no-cache",
    "Cache-Control": "no-cache",
    "Upgrade": "websocket",
}

url = "http://localhost:8001"
response = requests.get(url=url, headers=headers)
print(response.json())
```

and this was printed

scope: {'type': 'websocket', 'asgi': {'version': '3.0', 'spec_version': '2.1'}, 'http_version': '1.1', 'scheme': 'ws', 'method': 'GET', 'path': '/', 'raw_path': b'/', 'query_string': b'', 'client': ('127.0.0.1', 0), 'server': ('127.0.0.1', 80), 'headers': ((b'host', b'localhost:8001'), (b'user-agent', b'Mozilla/5.0 (X11; Linux x86_64; rv:134.0) Gecko/20100101 Firefox/134.0'), (b'accept-encoding', b'gzip, deflate'), (b'accept', b'*/*'), (b'connection', b'Upgrade'), (b'accept-language', b'en-GB,en;q=0.5'), (b'sec-websocket-version', b'13'), (b'origin', b'null'), (b'sec-websocket-extensions', b'permessage-deflate'), (b'sec-websocket-key', b'ckhEnA8kgUUzH43nqi2K8A=='), (b'dnt', b'1'), (b'pragma', b'no-cache'), (b'cache-control', b'no-cache'), (b'upgrade', b'websocket')), 'state': {}}



scope: {'type': 'http', 'asgi': {'version': '3.0', 'spec_version': '2.1'}, 'http_version': '1.1', 'scheme': 'http', 'method': 'GET', 'path': '/', 'raw_path': b'/', 'query_string': b'', 'client': ('127.0.0.1', 0), 'server': ('127.0.0.1', 80), 'headers': ((b'host', b'localhost:8001'), (b'user-agent', b'Mozilla/5.0 (X11; Linux x86_64; rv:134.0) Gecko/20100101 Firefox/134.0'), (b'accept-encoding', b'gzip, deflate'), (b'accept', b'*/*'), (b'connection', b'keep-alive, Upgrade'), (b'accept-language', b'en-GB,en;q=0.5'), (b'sec-websocket-version', b'13'), (b'origin', b'null'), (b'sec-websocket-extensions', b'permessage-deflate'), (b'sec-websocket-key', b'ckhEnA8kgUUzH43nqi2K8A=='), (b'dnt', b'1'), (b'pragma', b'no-cache'), (b'cache-control', b'no-cache'), (b'upgrade', b'websocket')), 'state': {}}

**Comment by ac000** [comment]
> The difference I see in the request from firefox and chromium is in the field ("Connection")

> Chromium: Connection: Upgrade
> Firefox: Connection: keep-alive, Upgrade

Yeah, I did notice that, next thing to investigate...

> Also, For case where unit is able to identify it is websocket upgrade request. then I guess changing line at this places will make unit strictly asgi conformant in the nxt_py_asgi_create_http_scope function, instead of setting method for both type(http and ws). It can be set conditionaly
by checking the type value.

> SET_ITEM(scope, method, v)

Heh, yeah I had tried that, but with Firefox... and it seems we never even hit the Python stuff with it...

I.e.

When Chromium connects we hit the following functions and all is good.

```
nxt_h1p_header_process: 
nxt_h1p_connection: 
nxt_router_prepare_msg: 
nxt_router_prepare_msg: Done websocket handshake
nxt_py_asgi_create_http_scope: 
```

When Firefox tries we get

```
nxt_h1p_header_process: 
nxt_h1p_connection: 
nxt_router_prepare_msg: 
nxt_py_asgi_create_http_scope: 
```

You'll notice that we only call into `nxt_router_prepare_msg()` once, so we never even get to Python.

So that does suggest an issue with the `Connection` header...

Now that I have something that does basically work with Unit Websockets, let me try not setting the method again...

Bingo!

```
(gdb) b nxt_py_asgi_create_http_scope
Breakpoint 1 at 0x7fad10c6bf1d: file src/python/nxt_python_asgi.c, line 652.
(gdb) c
Continuing.

Breakpoint 1, nxt_py_asgi_create_http_scope (req=0x23a6ae0, 
    app_target=0x2221058) at src/python/nxt_python_asgi.c:652
652	    printf("%s: \n", __func__);
(gdb) n
665	    v = NULL;
(gdb) 
666	    headers = NULL;
(gdb) 
668	    r = req->request;
(gdb) 
670	    if (r->websocket_handshake) {
(gdb) p r->websocket_handshake 
$1 = 1 '\001'
```

OK, now we know that we're a WebSocket... and with the following patch

```diff
diff --git a/src/python/nxt_python_asgi.c b/src/python/nxt_python_asgi.c
index 8f300b53..8f950931 100644
--- ./src/python/nxt_python_asgi.c
+++ ./src/python/nxt_python_asgi.c
@ -693,15 +695,17 @@ nxt_py_asgi_create_http_scope(nxt_unit_request_info_t *req,
                                               : nxt_py_1_0_str)
     SET_ITEM(scope, scheme, scheme)
 
-    v = PyString_FromStringAndSize(nxt_unit_sptr_get(&r->method),
-                                   r->method_length);
-    if (nxt_slow_path(v == NULL)) {
-        nxt_unit_req_alert(req, "Python failed to create 'method' string");
-        goto fail;
-    }
+    if (!r->websocket_handshake) {
+        v = PyString_FromStringAndSize(nxt_unit_sptr_get(&r->method),
+                                       r->method_length);
+        if (nxt_slow_path(v == NULL)) {
+            nxt_unit_req_alert(req, "Python failed to create 'method' string");
+            goto fail;
+        }
 
-    SET_ITEM(scope, method, v)
-    Py_DECREF(v);
+        SET_ITEM(scope, method, v)
+        Py_DECREF(v);
+    }
 
     v = PyUnicode_DecodeUTF8(nxt_unit_sptr_get(&r->path), r->path_length,
                              "replace");
```

The Litestar app now works (I think, Chromium at least tells me the Websocket is connected).

**Comment by ac000** [comment]
Regarding the Firefox connection issue, looks we're only expecting a single value in that field...

```c
static nxt_int_t                                                                
nxt_h1p_connection(void *ctx, nxt_http_field_t *field, uintptr_t data)          
{                                                                               
    nxt_http_request_t  *r; 
                                                                                
    r = ctx;                                                                    
    field->hopbyhop = 1;                                                        
                                                                                
    if (field->value_length == 5                                                
        && nxt_memcasecmp(field->value, "close", 5) == 0)                       
    {                                                                           
        r->proto.h1->keepalive = 0;                                             
                                                                                
    } else if (field->value_length == 10                                        
               && nxt_memcasecmp(field->value, "keep-alive", 10) == 0)          
    {                                                                           
        r->proto.h1->keepalive = 1;                                             
                                                                                
    } else if (field->value_length == 7                                         
               && nxt_memcasecmp(field->value, "upgrade", 7) == 0)              
    {                                                                           
        r->proto.h1->connection_upgrade = 1;                                    
    }                                                                           
                                                                                
    return NXT_OK;                                                              
} 
```

Still, that should be easy enough to fix... though maybe I need to check if `Connection: keep-alive, Upgrade` is supposed to have some special meaning...

But this is a separate issue to this one...

**Comment by gourav-kandoria** [comment]
> Regarding the Firefox connection issue, looks like we're only expecting a single value in that field...
> 
> static nxt_int_t                                                                
> nxt_h1p_connection(void *ctx, nxt_http_field_t *field, uintptr_t data)          
> {                                                                               
>     nxt_http_request_t  *r; 
>                                                                                 
>     r = ctx;                                                                    
>     field->hopbyhop = 1;                                                        
>                                                                                 
>     if (field->value_length == 5                                                
>         && nxt_memcasecmp(field->value, "close", 5) == 0)                       
>     {                                                                           
>         r->proto.h1->keepalive = 0;                                             
>                                                                                 
>     } else if (field->value_length == 10                                        
>                && nxt_memcasecmp(field->value, "keep-alive", 10) == 0)          
>     {                                                                           
>         r->proto.h1->keepalive = 1;                                             
>                                                                                 
>     } else if (field->value_length == 7                                         
>                && nxt_memcasecmp(field->value, "upgrade", 7) == 0)              
>     {                                                                           
>         r->proto.h1->connection_upgrade = 1;                                    
>     }                                                                           
>                                                                                 
>     return NXT_OK;                                                              
> } 
> 
> Still, that should be easy enough to fix... though maybe I need to check if `Connection: keep-alive, Upgrade` is supposed to have some special meaning...
> 
> But this is a separate issue to this one...

Understood. So handling multiple valid values correctly here would also make this work correclty

**Comment by ac000** [comment]
> Understood. So handling multiple valid values correctly here would also make firefox to open ws connection open correctly.

Indeed.

**Comment by FixFlare** [comment]
@ac000 and @gourav-kandoria 

I just built Unit from source on my test server. It's showing version 1.35 now and I can confirm everything is working as expected with an unmodified install of Litestar 2.14.0.

Thank you for your efforts on this and the quick fix. It's going to make my deployments so much easier now.

**Comment by ac000** [comment]
Thanks for confirming.
